File: Week 3/2.For_loop_variations.java
/*
 * WEEK 3 - LOOPS & NUMBER THEORY
 * Topic: Variations of For Loop
 * File: 2.For_loop_variations.java
 *
 * CONCEPT:
 * The three expressions inside a for loop (initialization, condition, update)
 * are all optional. This file demonstrates various combinations of omitting
 * one or more parts, including infinite loops and multi-variable loops.
 *
 * KEY POINTS / ALGORITHM:
 * 1. Initialization can be moved before the loop.
 * 2. Update can be placed inside the loop body.
 * 3. Omitting the condition creates an infinite loop.
 * 4. Omitting all three parts also creates an infinite loop: for(;;).
 * 5. Multiple variables can be initialized, tested, and updated using commas.
 * 6. Multiple conditions must be combined with logical operators (&&, ||).
 *
 * Time Complexity: O(n) for finite loops, infinite for loops without termination
 * Space Complexity: O(1) - only loop control variables used
 */


// Example code 1: Initialization part removed - declared before the loop
public static void main(String[] args) {
      int i = 0;                        // initialization done outside for(...)
      for( ; i < 3; i++) {              // only condition and update remain
          System.out.println(i);
      }
    }

/*
Output:
0
1
2

*/

//Example code 2: Updation part removed

public static void main(String[] args) {
      for(int i = 0; i < 3; ) {
          System.out.println(i);
          i++;
      }
    }

/*
Output:
0
1
2
*/


// Example code 3: Condition expression removed , thus making our loop infinite –

public static void main(String[] args) {
      for(int i = 0; ; i++) {
          System.out.println(i);
      }
  }


//Example code 4:
//We can remove all the three expression, thus forming an infinite loop

public static void main(String[] args) {
for( ; ; ) {
      System.out.print("Inside for loop");
    }
}

/*
● Multiple statements inside for loop
We can initialize multiple variables, have multiple conditions and multiple update
statements inside a for loop. We can separate multiple statements using
comma, but not for conditions. They need to be combined using logical
operators.


Example code:

*/
public static void main(String[] args) {
      for(int i = 0, j = 4; i < 5 && j >= 0; i++, j--) {
          System.out.println(i + " " + j);
      }
    }
/*Output:
0 4
1 3
2 2
3 1
4 0
*/

File: Week 3/cpp/2.for_loop_variations.cpp
/*
 * WEEK 3 - C++ LOOPS & NUMBER THEORY
 * Topic: For Loop Variations
 * File: 2.for_loop_variations.cpp
 *
 * CONCEPT:
 *  All three parts of `for (init; cond; update)` are optional.
 *  Multiple init/update statements separated by commas; multiple
 *  conditions joined with `&&` / `||`.
 *
 * KEY POINTS:
 *  - `for (;;) { ... }` is an infinite loop (canonical).
 *  - `for (int i=0, j=4; i<5 && j>=0; ++i, --j) { ... }` -- two counters.
 *  - `for (int i = init(); i; --i)` -- truthy condition uses the int value directly.
 *
 * DRY RUN:
 *  Two-counter loop: (0,4), (1,3), (2,2), (3,1), (4,0)
 */

#include <iostream>

int main() {
    // 1) Init removed
    int i = 0;
    for (; i < 3; ++i) std::cout << i << ' ';
    std::cout << '\n';

    // 2) Update removed (done inside body)
    for (int j = 0; j < 3; ) {
        std::cout << j << ' ';
        ++j;
    }
    std::cout << '\n';

    // 3) Condition removed -- infinite. Capped here.
    int safety = 0;
    for (int k = 0; ; ++k) {
        std::cout << k << ' ';
        if (++safety >= 3) break;
    }
    std::cout << '\n';

    // 4) Two counters with comma operator
    for (int a = 0, b = 4; a < 5 && b >= 0; ++a, --b) {
        std::cout << a << ' ' << b << '\n';
    }

    // 5) for(;;) -- canonical infinite loop
    int n = 0;
    for (;;) {
        if (++n >= 3) { std::cout << "stopped at n=" << n << '\n'; break; }
    }
    return 0;
}

/*
 * NOTES:
 *  - The comma operator `,` evaluates both expressions and yields the right one.
 *  - Multi-variable for headers can declare ONLY ONE TYPE -- so all loop vars
 *    must share a type. Use auto with structured bindings if needed.
 *  - For mixing types use a struct / pair / tuple in the init.
 */
